dbms/src/Storages/MergeTree/KeyCondition.cpp

```diff
@@ -444,6 +444,45 @@ void KeyCondition::traverseAST(const ASTPtr & node, const Context & context, Blo
                   */
                 if (i != 0 || element.function == RPNElement::FUNCTION_NOT)
                     rpn.emplace_back(std::move(element));
+
+                /* Combine statements like (x >= a) and (x <= b) into x in range [a, b] */
+                if (rpn.size() >= 3 && rpn.back().function == RPNElement::FUNCTION_AND)
+                {
+                    const size_t to_modify = rpn.size() - 3;
+                    const auto rhs = rpn[to_modify + 1];
+                    const auto lhs = rpn[to_modify];
```

Comment:
We can use references here.

Also, let's create a merged element on stack and then put it in `rpn` if the merge succeeds -- when modifying in place, we can leave some rubbish, especially if some field is added in the future.